nomic/data_operations.py

```diff
@@ -1,42 +1,602 @@
+        if queries is not None:
+            if not isinstance(queries, np.ndarray):
+                raise Exception("`queries` must be an instance of np.array.")
+            if queries.shape[0] > max_queries:
+                raise Exception(f"Max vectors per query is {max_queries}. You sent {queries.shape[0]}.")
+
+        if queries is not None:
+            if queries.ndim != 2:
+                raise ValueError(
+                    'Expected a 2 dimensional array. If you have a single query, we expect an array of shape (1, d).'
+                )
+
+            bytesio = io.BytesIO()
+            np.save(bytesio, queries)
+
+        if queries is not None:
+            response = requests.post(
+                self.projection.project.atlas_api_path + "/v1/project/data/get/nearest_neighbors/by_embedding",
+                headers=self.projection.project.header,
+                json={
+                    'atlas_index_id': self.projection.atlas_index_id,
+                    'queries': base64.b64encode(bytesio.getvalue()).decode('utf-8'),
+                    'k': k,
+                },
+            )
+        else:
+            response = requests.post(
+                self.projection.project.atlas_api_path + "/v1/project/data/get/nearest_neighbors/by_id",
+                headers=self.projection.project.header,
+                json={'atlas_index_id': self.projection.atlas_index_id, 'datum_ids': ids, 'k': k},
+            )
+
+        if response.status_code == 500:
+            raise Exception('Cannot perform vector search on your map at this time. Try again later.')
+
+        if response.status_code != 200:
+            raise Exception(response.text)
+
+        response = response.json()
+
+        return response['neighbors'], response['distances']
+
+    def get_embedding_iterator(self) -> Iterable[Tuple[str, str]]:
+        '''
+        Iterate through embeddings of your datums.
+
+        Returns:
+            A iterable mapping datum ids to their embeddings.
+
+        '''
+
+        if self.project.is_locked:
+            raise Exception('Project is locked! Please wait until the project is unlocked to download embeddings')
+
+        offset = 0
+        limit = EMBEDDING_PAGINATION_LIMIT
+        while True:
+            response = requests.get(
+                self.atlas_api_path
+                + f"/v1/project/data/get/embedding/{self.project.id}/{self.projection.atlas_index_id}/{offset}/{limit}",
+                headers=self.header,
+            )
+            if response.status_code != 200:
+                raise Exception(response.text)
+
+            content = response.json()
+            if len(content['datum_ids']) == 0:
+                break
+            offset += len(content['datum_ids'])
+
+            yield content['datum_ids'], content['embeddings']
+
+    def download_embeddings(self, save_directory: str, num_workers: int = 10) -> bool:
+        '''
+        Downloads shards of arrow tables that map
```

Comment:
nit: is this complete?

Reply:
all of AtlasMapEmbeddings is not fully complete yet. it is blocked by ben implementing logic on the atlas cloud backend for writing embeddings more efficiently for export.

Once we have an efficient method pull out ambient embeddings, accessing with .ambient will be possible.

For now the existing logic where we download all embeddings to a directory is here.